src/main/kotlin/de/bund/digitalservice/useid/credentials/CredentialService.kt

```diff
@@ -0,0 +1,135 @@
+package de.bund.digitalservice.useid.credentials
+
+import com.yubico.webauthn.AssertionRequest
+import com.yubico.webauthn.FinishAssertionOptions
+import com.yubico.webauthn.FinishRegistrationOptions
+import com.yubico.webauthn.RelyingParty
+import com.yubico.webauthn.StartAssertionOptions
+import com.yubico.webauthn.StartRegistrationOptions
+import com.yubico.webauthn.data.AuthenticatorSelectionCriteria
+import com.yubico.webauthn.data.ByteArray
+import com.yubico.webauthn.data.PublicKeyCredential
+import com.yubico.webauthn.data.ResidentKeyRequirement
+import com.yubico.webauthn.data.UserIdentity
+import com.yubico.webauthn.exception.AssertionFailedException
+import mu.KotlinLogging
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.stereotype.Service
+import java.security.SecureRandom
+import java.util.UUID
+import kotlin.jvm.optionals.getOrDefault
+
+@Service
+@ConditionalOnProperty(name = ["features.desktop-solution-enabled"], havingValue = "true")
+class CredentialService(
+    private val credentialMockDatasource: CredentialMockDatasource,
+    private val relyingParty: RelyingParty,
+) {
+
+    private val log = KotlinLogging.logger {}
+
+    fun startRegistration(widgetSessionId: UUID, refreshAddress: String): Credential {
+        val user = buildUser(widgetSessionId)
+        val startOptions = buildStartRegistrationOptions(user)
+
+        val pkcCreationOptions = relyingParty.startRegistration(startOptions)
+
+        val credential = credentialMockDatasource.save(
+            Credential(
+                UUID.randomUUID(),
+                widgetSessionId,
+                user.name,
+                user.id.base64,
+                refreshAddress,
+                pkcCreationOptions,
+            ),
+        )
+        log.info("Created new user credential. credentialId=${UUID.randomUUID()}")
+        return credential
+    }
+
+    private fun buildUser(widgetSessionId: UUID): UserIdentity {
+        val username = widgetSessionId.toString()
+        return UserIdentity.builder()
+            .name(username)
+            .displayName(username)
+            .id(generateUserId())
+            .build()
+    }
+
+    private fun buildStartRegistrationOptions(user: UserIdentity): StartRegistrationOptions? {
```

Comment:
Why does this have an optional return type?

Reply:
Good catch. Auto extracted method. Removed.